R/bookmark-state.R

```diff
@@ -0,0 +1,960 @@
+
+
+      # Error if multiple '_inputs_' or '_values_'. This is needed because
+      # strsplit won't add an entry if the search pattern is at the end of a
+      # string.
+      if (length(gregexpr("(^|&)_inputs_(&|$)", queryString)[[1]]) > 1)
+        stop("Invalid state string: more than one '_inputs_' found")
+      if (length(gregexpr("(^|&)_values_(&|$)", queryString)[[1]]) > 1)
+        stop("Invalid state string: more than one '_values_' found")
+
+      # Look for _inputs_ and store following content in inputStr
+      splitStr <- strsplit(queryString, "(^|&)_inputs_(&|$)")[[1]]
+      if (length(splitStr) == 2) {
+        inputStr <- splitStr[2]
+        # Remove any _values_ (and content after _values_) that may come after
+        # _inputs_
+        inputStr <- strsplit(inputStr, "(^|&)_values_(&|$)")[[1]][1]
+
+      } else {
+        inputStr <- ""
+      }
+
+      # Look for _values_ and store following content in valueStr
+      splitStr <- strsplit(queryString, "(^|&)_values_(&|$)")[[1]]
+      if (length(splitStr) == 2) {
+        valueStr <- splitStr[2]
+        # Remove any _inputs_ (and content after _inputs_) that may come after
+        # _values_
+        valueStr <- strsplit(valueStr, "(^|&)_inputs_(&|$)")[[1]][1]
+
+      } else {
+        valueStr <- ""
+      }
+
+
+      inputs <- parseQueryString(inputStr, nested = TRUE)
+      values <- parseQueryString(valueStr, nested = TRUE)
+
+      valuesFromJSON <- function(vals) {
+        mapply(names(vals), vals, SIMPLIFY = FALSE,
+          FUN = function(name, value) {
+            tryCatch(
+              jsonlite::fromJSON(value),
+              error = function(e) {
+                stop("Failed to parse URL parameter \"", name, "\"")
+              }
+            )
+          }
+        )
+      }
+
+      inputs <- valuesFromJSON(inputs)
+      self$input <- RestoreInputSet$new(inputs)
+
+      values <- valuesFromJSON(values)
+      self$values <- list2env(values, self$values)
+    }
+  )
+)
+
+
+# Restore input set. This is basically a key-value store, except for one
+# important difference: When the user `get()`s a value, the value is marked as
+# pending; when `flushPending()` is called, those pending values are marked as
+# used. When a value is marked as used, `get()` will not return it, unless
+# called with `force=TRUE`. This is to make sure that a particular value can be
+# restored only within a single call to `withRestoreContext()`. Without this, if
+# a value is restored in a dynamic UI, it could completely prevent any other
+# (non- restored) kvalue from being used.
+RestoreInputSet <- R6Class("RestoreInputSet",
+  private = list(
+    values = NULL,
+    pending = character(0),
+    used = character(0)     # Names of values which have been used
+  ),
+
+  public = list(
+    initialize = function(values) {
+      private$values <- new.env(parent = emptyenv())
+      list2env(values, private$values)
```

Comment:
Nope, the result is exactly the same. I'll use your version because it's slightly simpler.